src/arkode/arkode_mristep.c

```diff
@@ -1285,43 +1476,94 @@ int mriStep_FullRHS(ARKodeMem ark_mem, sunrealtype t, N_Vector y, N_Vector f,
-      retval = mriStepInnerStepper_FullRhs(step_mem->stepper, t, y, f,
-                                           ARK_FULLRHS_OTHER);
-      if (retval != ARK_SUCCESS)
+        /* copy the implicit component */
+        if (step_mem->implicit_rhs)
+        {
+          N_VScale(ONE, step_mem->Fsi[sa_stage], step_mem->Fsi[0]);
+        }
+      }
+      else
       {
-        arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
-                        MSG_ARK_RHSFUNC_FAILED, t);
-        return (ARK_RHSFUNC_FAIL);
+        /* compute the explicit component */
+        if (step_mem->explicit_rhs)
+        {
+          retval = step_mem->fse(t, y, step_mem->Fse[0], ark_mem->user_data);
+          step_mem->nfse++;
+          step_mem->fse_is_current = SUNTRUE;
+          if (retval != 0)
+          {
+            arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__,
+                            __FILE__, MSG_ARK_RHSFUNC_FAILED, t);
+            return (ARK_RHSFUNC_FAIL);
+          }
+
+          /* Add external forcing, as appropriate */
+          if (step_mem->expforcing)
+          {
+            step_mem->cvals[0] = ONE;
+            step_mem->Xvecs[0] = step_mem->Fse[0];
+            nvec               = 1;
+            mriStep_ApplyForcing(step_mem, t, ONE, &nvec);
+            N_VLinearCombination(nvec, step_mem->cvals, step_mem->Xvecs,
+                                 step_mem->Fse[0]);
+          }
+        }
+
+        /* compute the implicit component */
+        if (step_mem->implicit_rhs)
+        {
+          retval = step_mem->fsi(t, y, step_mem->Fsi[0], ark_mem->user_data);
+          step_mem->nfsi++;
+          step_mem->fsi_is_current = SUNTRUE;
+          if (retval != 0)
+          {
+            arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__,
+                            __FILE__, MSG_ARK_RHSFUNC_FAILED, t);
+            return (ARK_RHSFUNC_FAIL);
+          }
+
+          /* Add external forcing, as appropriate */
+          if (step_mem->impforcing)
+          {
+            step_mem->cvals[0] = ONE;
+            step_mem->Xvecs[0] = step_mem->Fsi[0];
+            nvec               = 1;
+            mriStep_ApplyForcing(step_mem, t, ONE, &nvec);
+            N_VLinearCombination(nvec, step_mem->cvals, step_mem->Xvecs,
+                                 step_mem->Fsi[0]);
+          }
+        }
       }
     }
 
+    /* compute the fast component (force new RHS computation) */
+    retval = mriStepInnerStepper_FullRhs(step_mem->stepper, t, y, f,
+                                         ARK_FULLRHS_OTHER);
+    if (retval != ARK_SUCCESS)
+    {
+      arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                      MSG_ARK_RHSFUNC_FAILED, t);
+      return (ARK_RHSFUNC_FAIL);
+    }
+
     /* combine RHS vectors into output */
     if (step_mem->explicit_rhs && step_mem->implicit_rhs)
```

Comment:
What do you thin about if inside `mriStep_TakeStepMRISR`, I change
```
    /* combine both RHS into Fse for ImEx problems */
    if (step_mem->implicit_rhs && step_mem->explicit_rhs)
    {
      N_VLinearSum(ONE, step_mem->Fse[0], ONE, step_mem->Fsi[0],
                   step_mem->Fse[0]);
    }
```
to
```
    /* combine both RHS into Fse for ImEx problems */
    if (step_mem->implicit_rhs && step_mem->explicit_rhs)
    {
      N_VLinearSum(ONE, step_mem->Fse[0], ONE, step_mem->Fsi[0],
                   step_mem->Fse[0]);
      N_VConst(ZERO, step_mem->Fsi[0]);
    }
```
This would ensure that the sum never double-counts anything, without adding extra complication to test for stepper/method types?

Reply:
Seems like a much simpler solution. It would be good to include a comment to clarify why we're doing this for future reference.